src/clients/Deluge.ts

```diff
@@ -0,0 +1,236 @@
+				"you need to define a password in the delugeWebUrl. (eg: http://:<PASSWORD>@localhost:8112)"
+			);
+		}
+		const result = await this.authenticate();
+		if (!result) {
+			logger.verbose({
+				label: Label.DELUGE,
+				message: "failed to authenticate the deluge daemon.",
+			});
+			throw new CrossSeedError(
+				`failed to authenticate the deluge daemon.`
+			);
+		} else {
+			this.enLabeled = this.labelEnabled();
+		}
+	}
+
+	/**
+	 * connects and authenticates to the webui
+	 */
+	private async authenticate(): Promise<boolean> {
+		if (!this.loggedIn) {
+			let response = await this.call({
+				params: [this.delugeWebUrl.password],
+				method: "auth.login",
+			});
+			if (response === null) {
+				throw new CrossSeedError(
+					`Could not reach deluge: ${this.delugeWebUrl.origin}`
+				);
+			}
+			if (
+				response &&
+				response.headers &&
+				response["data"]["result"] &&
+				response.headers["set-cookie"]
+			) {
+				this.delugeCookie =
+					response.headers["set-cookie"][0].split(";")[0];
+				this.loggedIn = true;
+			}
+		} else {
+			let response = await this.call({
+				params: [this.delugeCookie],
+				method: "auth.check_session",
+			});
+			this.loggedIn = !!response;
+		}
+		return this.loggedIn;
+	}
+
+	/**
+	 * authenticates and sends JSON-RPC calls to deluge
+	 */
+	private async sendCall(method: string, params: any[]) {
+		let response;
+		if (!this.loggedIn) {
+			await this.authenticate();
+			response = await this.call({
+				params,
+				method,
+			});
+		} else
+			response = await this.call({
+				params: params,
+				method,
+			});
+
+		return response && response.data && response.data.error === null
+			? response.data
+			: response.data.error;
+	}
+
+	private async call(body: any): Promise<AxiosResponse | null> {
+		return new Promise((resolve) => {
+			body.id = ++this.msgId;
+			if (this.msgId > 1024) this.msgId = 0;
+			axios
+				.post(
+					this.delugeWebUrl.origin + this.delugeWebUrl.pathname,
+					body,
+					{
+						headers: {
+							"Content-Type": "application/json",
+							Cookie: this.delugeCookie,
+						},
+					}
+				)
+				.then((data) => {
+					resolve(data);
+				})
+				.catch(() => resolve(null));
+		});
+	}
+
+	/**
+	 * lists plugins and adds/sets labels
+	 * returns true if successful.
+	 */
+	private async labelEnabled() {
+		const enabledLabels = await this.sendCall(
+			"core.get_enabled_plugins",
+			[]
+		);
+		return enabledLabels["result"].includes("Label");
+	}
+	private async setLabel(torrenthash: string) {
+		if (this.enLabeled) {
+			let setResult = await this.sendCall("label.set_torrent", [
+				torrenthash,
+				this.delugeLabel,
+			]);
+			if (
+				setResult["message"] &&
+				setResult["message"].includes("Unknown Label")
+			) {
+				await this.sendCall("label.add", [this.delugeLabel]);
+				await this.sendCall("label.set_torrent", [
+					torrenthash,
+					this.delugeLabel,
+				]);
+			}
+			return true;
+		} else {
+			return false;
+		}
```

Comment:
do we need to return this boolean? might be able to get away with returning void

Reply:
I guess this is redundant. I had the idea to forcibly enable labels plugin, and I think that's where this typing originated from.